### PR TITLE
fix: respect heredoc shebang for ShellCheck dialect detection

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

- Docker BuildKit respects `#!/bin/bash` shebangs inside heredoc bodies (`RUN <<EOF`), but tally was ignoring them and always linting with the stage's default shell — causing false SC3xxx violations on valid bash code
- Detect heredoc shebangs via `shell.ShellFromShebang()` (wraps `mvdan.cc/sh/v3/fileutil.Shebang` + ksh fallback) and override the ShellCheck dialect
- Skip the synthetic `#!/bin/sh` prelude when the script carries its own shebang; suppress SC1128 since prelude exports push the shebang past line 1
- Enrich the semantic model with `HeredocShellOverrides` on `StageInfo` so other rules can query per-instruction shell overrides

## Test plan

- [x] Integration test `shellcheck-heredoc-shebang` — Dockerfile with `#!/bin/bash` heredoc using bash array syntax; expects zero violations
- [x] Validated the fixture Dockerfile builds successfully with `docker build --no-cache`
- [x] Unit tests for `ShellFromShebang()` (14 cases: all shell types, env-style, non-shell, edge cases)
- [x] Unit test for `HeredocShellOverride` population in semantic builder
- [x] All existing shellcheck tests pass (basic, sc1040, fallback-empty-heredoc)
- [x] Full suite: 3821 tests pass, 0 failures
- [x] `make lint` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Detect shebangs inside heredocs to determine the script shell and improve dialect handling.

* **Bug Fixes**
  * Suppress incorrect ShellCheck warnings for valid heredoc scripts that start with a shebang.

* **Tests**
  * Added unit and integration tests (including a new snapshot) covering heredoc shebang detection and per-RUN heredoc shell overrides.

### CODE REVIEW

File: internal/rules/shellcheck/shellcheck.go [hunk 1]
```diff
@@ -433,7 +433,7 @@ func (r *Rule) checkShellSnippet(
 
 	baseLoc := rules.NewLocationFromRanges(file, location)
 
-	prelude, _ := buildPrelude(dialect, knownEnv)
+	prelude, _ := buildPrelude(dialect, knownEnv, false)
 	script := prelude + snippet
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/shellcheck/shellcheck.go]: ✅ Addressed in 15da7c5. Thanks @coderabbitai\!
